test/scheduler/tainttoleration_integration_test.go

```diff
@@ -0,0 +1,702 @@
+			// Delete the provisional cluster.
+			ensureProvisionalClusterDeletion(newUnhealthyMemberClusterName)
+		})
+	})
+
+	// This is a serial test as adding a new member cluster may interrupt other test cases.
+	Context("pickAll, add a new healthy cluster with taint and matching toleration", Serial, Ordered, func() {
+		crpName := fmt.Sprintf(crpNameTemplate, GinkgoParallelProcess())
+		policySnapshotName := fmt.Sprintf(policySnapshotNameTemplate, crpName, 1)
+		// Prepare a new cluster to avoid interrupting other concurrently running test cases.
+		newUnhealthyMemberClusterName := fmt.Sprintf(provisionalClusterNameTemplate, GinkgoParallelProcess())
+		updatedHealthyClusters := healthyClusters
+		updatedHealthyClusters = append(updatedHealthyClusters, newUnhealthyMemberClusterName)
+		// Copy the map to avoid interrupting other concurrently running test cases.
+		updatedZeroScoreByCluster := make(map[string]*placementv1beta1.ClusterScore)
+		for k, v := range zeroScoreByCluster {
+			updatedZeroScoreByCluster[k] = v
+		}
+		updatedZeroScoreByCluster[newUnhealthyMemberClusterName] = &zeroScore
+
+		BeforeAll(func() {
+			// Ensure that no bindings have been created so far.
+			noBindingsCreatedActual := noBindingsCreatedForCRPActual(crpName)
+			Consistently(noBindingsCreatedActual, consistentlyDuration, consistentlyInterval).Should(Succeed(), "Some bindings have been created unexpectedly")
+
+			// Create a CRP with no scheduling policy specified, along with its associated policy snapshot, and toleration for new cluster.
+			createNilSchedulingPolicyCRPWithPolicySnapshot(crpName, policySnapshotName, buildTolerations([]string{newUnhealthyMemberClusterName}))
+
+			// Create a new member cluster.
+			createMemberCluster(newUnhealthyMemberClusterName, buildTaints([]string{newUnhealthyMemberClusterName}))
+
+			// Mark this cluster as healthy.
+			markClusterAsHealthy(newUnhealthyMemberClusterName)
+		})
+
+		It("should create scheduled bindings for the newly recovered cluster with tolerated taint", func() {
+			scheduledBindingsCreatedActual := scheduledBindingsCreatedOrUpdatedForClustersActual(updatedHealthyClusters, updatedZeroScoreByCluster, crpName, policySnapshotName)
+			Eventually(scheduledBindingsCreatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to create the expected set of bindings")
+			Consistently(scheduledBindingsCreatedActual, consistentlyDuration, consistentlyInterval).Should(Succeed(), "Failed to create the expected set of bindings")
+		})
+
+		AfterAll(func() {
+			// Delete the CRP.
+			ensureCRPAndAllRelatedResourcesDeletion(crpName)
+			// Delete the provisional cluster.
+			ensureProvisionalClusterDeletion(newUnhealthyMemberClusterName)
+		})
+	})
+
+	// This is a serial test as adding a new member cluster may interrupt other test cases.
+	Context("pickN with required topology spread constraints, add new cluster with taint, upscaling doesn't pick new cluster", Serial, Ordered, func() {
```

Comment:
can we make the test simple, focusing on the taint feature by removing the topology spread?

The topology spread has been already covered.

Reply:
we add a complicated policy in the e2e tests.